### PR TITLE
Feature owner-reference

### PR DESCRIPTION
**Pull request by abhi16394**
# Description
Adding owner-reference support to csm-operator.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
|https://github.com/dell/csm/issues/255 |

# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [x] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have maintained backward compatibility

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

### CODE REVIEW

**Review by rensyct — CHANGES_REQUESTED** [review]
Please attach screenshots of dev sanity tests

**Comment by abhi16394** [comment]
image

**Comment by abhi16394** [comment]
image

**Review by yifeijin — APPROVED** [review]

**Review by jooseppi-luna — APPROVED** [review]

File: controllers/csm_controller.go [hunk 1]
```diff
@@ -163,6 +163,14 @@ func (r *ContainerStorageModuleReconciler) Reconcile(ctx context.Context, req ct
 		ConfigDirectory: r.Config.ConfigDirectory,
 	}
 
+	// perform prechecks
```
Comment (medegw01) [inline controllers/csm_controller.go:166]: Why are we changing the order? Is there a reason we want to do a precheck before checking if a file should be deleted?

One problem we will most likely face is when we deploy a CR which fails precheck.  If we try to delete that bad CR, the operator will prevent this as the precheck will  still be failing.  Nonetheless, I don't see why we need to run precheck for CR that is already scheduled for deletion.

I'll recommend we at least move this precheck to after checking if the CR is being deleted.

**Review by medegw01 — CHANGES_REQUESTED** [review]

File: controllers/csm_controller.go (hunk 1 above)
Comment (medegw01) [reply controllers/csm_controller.go:166]: Or we can leave owner ref check and driver/module pre-req checks in the Precheck function and call the precheck function after checking for deletion.

File: controllers/csm_controller.go [hunk 2]
```diff
@@ -728,6 +730,29 @@ func (r *ContainerStorageModuleReconciler) PreChecks(ctx context.Context, cr *cs
...
+				if m.Name == cr.Name {
+					log.Infow("Owner reference is found and matches")
+					break
+				} else {
+					return fmt.Errorf("Required Owner reference not found. Please re-install driver ")
+				}
+			}
+
+		} else {
+			return fmt.Errorf("Owner reference not found. Please re-install driver")
```
Comment (medegw01) [inline controllers/csm_controller.go:751]: Please see my other comment. The customer will be stuck here. When they try to delete the CR and deploy a new driver, the operator will prevent the custom resource deletion as the finalizer will still be set.

File: controllers/csm_controller.go (hunk 1 above)
Comment (abhi16394) [reply controllers/csm_controller.go:166]: We cannot call Prechecks after deletion. one of the reasons we are adding owner reference is to check if we have an existing installation of driver that was previously deployed by operator with owner reference. if we delete the driver even before checking for owner reference, it defeats the purpose and i wont be able to check owner reference if we delete the driver.
if we have a driver which fails other prechecks and doesn't have a owner reference then the driver has to be manually deleted.

File: controllers/csm_controller.go (hunk 2 above)
Comment (abhi16394) [reply controllers/csm_controller.go:751]: if we have a driver which doesn't have a owner reference, then we need to manually delete the driver by taking out finalizer and reinstalling the driver again. csm cannot help with deleting the driver if it doesn't recognize owner reference.

**Review by medegw01 — APPROVED** [review]

**Review by medegw01 — APPROVED** [review]

**Review by jooseppi-luna — APPROVED** [review]

**Comment by abhi16394** [comment]
Closing and opening a new Pull request to rebase from main branch

**Pull request closed without merging**